### PR TITLE
[FLINK-12248][table-planner-blink] Support e2e over window in blink batch

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Supporting all major over window operations

## Verifying this change

UT

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java [hunk 1]
```diff
@@ -67,6 +67,16 @@ public RexNodeConverter(RelBuilder relBuilder) {
 
 	@Override
 	public RexNode visitCall(CallExpression call) {
+		if (call.getFunctionDefinition().equals(BuiltInFunctionDefinitions.CAST)) {
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: Why not dealing with this inside `visitScalarFunc`?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/BoundComparatorCodeGenerator.scala [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    config: TableConfig,
+    inType: RowType,
+    bound: Any,
+    key: Int = -1,
+    keyType: InternalType = null,
+    keyOrder: Boolean = true,
+    isLowerBound: Boolean = true) {
+
+  def generateBoundComparator(name: String): GeneratedRecordComparator = {
+
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/BoundComparatorCodeGenerator.scala]: blank line

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/BoundComparatorCodeGenerator.scala [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+         return -1;
+       }
+     """.stripMargin
+  }
+}
+
+/**
+  * RANGE allow the compound ORDER BY and the random type when the bound is current row.
+  */
+class MultiFieldRangeBoundComparatorCodeGenerator(
```
Comment (KurtYoung) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/codegen/BoundComparatorCodeGenerator.scala]: use a new file

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/RexNodeConverter.java]: OK, I will take code into `visitScalarFunc`.
The reason I haven't done this before is that `TypeLiteralExpression` can't be converted. I will do a little refactoring and move the code to `visitScalarFunc`.

**Comment by KurtYoung** [comment]
16:45:21.073 [ERROR] Tests run: 1024, Failures: 8, Errors: 0, Skipped: 0

**Comment by JingsongLi** [comment]
Hi @godfreyhe 
I think you can pay attention to the test error of `HashAggregateTest` and `SortAggregateTest` Test, which is originally wrong.
I fix it first.

**Pull request merged**